Decouple dev workflow from Jupyter, create hot debugging environment

Nking92 opened the pull request:
Prior to this project, the back end and front end code were both tightly coupled with the Jupyter notebook environment. This project will allow us to run a React app using webpack-dev-server, and run the APIs in a Flask server with debug mode. Both of these toolchains enable hot reloading of components as code is changed, which will significantly improve dev workflow.

Notes about the changes:

- To run the React app, run `npm start` in the project's root directory.
- To run Flask in dev mode, run `flask run` from the project's root directory. The APIs will be hosted at localhost:5050. This configuration is set up in the .flaskenv file.
- Webpack config has been modified to have 2 configs, 1 for dev and 1 for production. The dev config sets up the React hot loader and allows us to run webpack-dev-server to host the app at localhost:3000. 
- widget.css has been moved to the widget/ directory and will be compiled into the widget-build-dev.js file. This will allow the CSS hot loader to work. The style loader module doesn't seem to work with Jupyter, so widget.css will be copied to its original location as a build step as well.
- A VSCode debug configuration has been set up. Press F5 to launch the app in Chrome. The webpack server must be running for it to work. The VSCode debugger should attach and allow you to set breakpoints as usual.
- A bit of code has been moved around so it could be reused to set up models for the dev server.

ilmarinen commented on webpack.config.js:
```diff
@@ -11,6 +13,9 @@ module.exports = {
       $: "jquery",
       jQuery: "jquery"
+    }),
+    new MiniCssExtractPlugin({
```
Is this to generate the "widget.css" file within the resources folder? Since you moved the CSS to be included within the React app?

jlema replied:
+1 on the jupyter stylings - I think the .less files used by jupyter are here: https://github.com/jupyter/notebook/tree/master/notebook/static/notebook/less
In particular, this one seems of interest: https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/less/renderedhtml.less

ilmarinen commented on webpack.config.js:
```diff
@@ -29,15 +34,83 @@ module.exports = {
   }
 };
+
+const devConfig = {
```
Do we have a different build target here because we want to inject hot reload?

ilmarinen replied:
If that is the case, would it be possible to keep the same build target, but add the hot reload if we build in development mode?

Nking92 replied:
Yes, the general intent is to allow the hot loader to work. The problem I encountered was that the dev mode widget needs some extra stuff compiled into it. What we end up getting are 2 different build artifacts, much like how debug symbols are included in the build with other languages.

`widget-build-dev.js` has the following differences

- React hot loader compiled in
- CSS styles compiled into the JS (which enables hot loading of CSS)

It might be possible to parse command line arguments here and conditionally change up the build depending on that, but that seemed trickier than just having 2 different targets. The current setup will also allow you to test your changes in parallel with Jupyter and the React environment without any need to change up your workflow - just run `npx webpack` to build your latest changes into `widget-build.js`. There won't be any risk of `widget-build.js` and `widget-build-dev.js` conflicting with each other.

Let me know what you think - I'm open to trying to consolidate the build targets if you think it's worthwhile long term.

ilmarinen replied:
Ok this makes a lot of sense now.

Could we separate out the artifacts into different folders?

So have a top level folder called "development-artifacts" that collect all the build artifacts related to the development server. And drop the `widget-build.js` in there.

Nking92 replied:
That's a good idea. Initially I wasn't sure how to separate everything, but that should be doable now.

ilmarinen commented on backwardcompatibilityml/widget/app.py:
```diff
@@ -0,0 +1,10 @@
+
+app = Flask(__name__)
+CORS(app)
+init_dev_server(app)
```
You may need to remove the port specification from the .flaskenv file.

Nking92 replied:
I'll give that a try. It's probably better for all the initialization code to live here, then go through the CompatibilityAnalysis constructor so we use the same code path.

ilmarinen commented on backwardcompatibilityml/widget/resources/index.html:
```diff
@@ -0,0 +1,37 @@
+<!--
```
Is there a reason that we cannot re-use the existing widget.html file?

Nking92 replied:
widget.html is a Jinja template, which won't work with `webpack dev server` (unless there's some clever plugin that allows us to do that)

ilmarinen replied:
Ah, I understand.

Ok, so since we need to run both the webpack dev server and the Python fixture generation code anyway. Could we do this instead:

1. Have the Python fixture generation code, output the HTML (generated from the Junja2 template) and place that in the folder.

2. Have the Webpack dev server run and serve things up from that location.

Nking92 replied:
I think we can do that. Are you anticipating needing to inject more things into the template later on? (I noticed "data" is always empty)

ilmarinen replied:
Yeah the injection needs to be cleaned up. It is always possible. Part of the reason for suggesting this is so that we have as minimal differences as possible between how we load the widget in Jupyter vs the dev harness.

ilmarinen reviewed (changes requested):
@Nking92 Really nice work. Thank you.

Overall the approach is good. I'd like to recommend the following structural changes (I've left some comments to suggest how to do them).

1. The current approach works by creating a Flask development app that is separate and parallel to the app that is present within the CompatibilityAnalysis instance.

2. It would be great if we could just re-use this app, and have CompatibilityAnalysis run in development mode, where we can run it as a Flask server from the command line. This shouldn't be too hard. I have left some comments that should help with this.

Besides that, in a few places the development server is referenced as "localhost". It would be great if we could have it listen on "0.0.0.0" instead. I work on a local VM that takes on an IP, and I use the app  / widget by pointing a browser from my host computer to the VM to load the app. So localhost causes things to fail for me. I was able to get things to work by changing this, so you should be able to do it too.

jlema reviewed (changes requested):
+1 on the jupyter stylings - I think the .less files used by jupyter are here: https://github.com/jupyter/notebook/tree/master/notebook/static/notebook/less
In particular, this one seems of interest: https://github.com/jupyter/notebook/blob/master/notebook/static/notebook/less/renderedhtml.less

Nking92 commented on package.json:
```diff
@@ -7,6 +7,7 @@
     "plotly.js-dist": "^1.54.7",
     "react": "^16.7.0",
     "react-dom": "^16.7.0",
+    "react-hot-loader": "^4.13.0",
```
That should be moved to dev dependencies

ilmarinen commented on backwardcompatibilityml/widget/app.py:
```diff
@@ -0,0 +1,91 @@
+    optimizer_kwargs={"lr": learning_rate, "momentum": momentum},
+    NewErrorLossClass=bcloss.BCCrossEntropyLoss,
+    StrictImitationLossClass=bcloss.StrictImitationCrossEntropyLoss,
+    port=5050,
```
If you set this to port=5040 and leave the .flaskenv as is ... it all works with dev_mode set to False to skip your workaround.

ilmarinen commented on backwardcompatibilityml/widget/compatibility_analysis.py:
```diff
@@ -149,51 +214,19 @@ def __init__(self, folder_name, number_of_epochs, h1, h2, training_set, test_set
+            self.html_widget = HTML(html_string)
+            display(self.html_widget)
+
+        init_app_routes(self.flask_service, self.sweep_manager)
```
Note that in non-dev mode, this actually has to initialize the routes like so:

`init_app_routes(FlaskHelper.app, self.sweep_manager)`

I might have missed that before, if so I apologize.

Nking92 replied:
Good catch, thanks!

Nking92 commented:
The latest changes fix React hot loading and cleans up the code quite a bit.

- Use HtmlWebpackPlugin to generate index.html at runtime. Webpack will inject the bundled JS code for us. `public/index.html` will be used as a template.
- Removed the `widget-build-dev.js` hack. It's not necessary to output dev build artifacts - webpack dev server actually hosts them in memory. This also fixed React hot loading.
- Combined the 2 webpack configs into 1 by inspecting the `NODE_ENV` environment variable. `NODE_ENV` will be set to `production` when building with `npm run build`.
- `npm run build` is the new way to build instead of `npx webpack`. We need to do it that way to set `NODE_ENV` to production.
- Jupyter notebook styles copied into a style called `.dev-container`. I just inspected the styles with Chrome and extracted the relevant stuff. We won't need to import Jupyter's `.less` files this way.

ilmarinen commented on backwardcompatibilityml/widget/app.py:
```diff
@@ -0,0 +1,90 @@
+from backwardcompatibilityml.helpers import training
+from backwardcompatibilityml.helpers.models import LogisticRegression, MLPClassifier
+from rai_core_flask.flask_helper import FlaskHelper
+from .compatibility_analysis import CompatibilityAnalysis
```
Can you change this to an absolute import rather than relative.

ilmarinen replied:
And can we move this file to a different folder, so that it is outside of the widget? Basically so that it doesn't get packaged with the widget.

I'd suggest something like:
```
root_folder
|
|--development
     |
     |--app.py
     |--development-artifacts
     |--public
```

ilmarinen commented on backwardcompatibilityml/widget/compatibility_analysis.py:
```diff
@@ -149,51 +213,11 @@ def __init__(self, folder_name, number_of_epochs, h1, h2, training_set, test_set
-        @FlaskHelper.app.route("/api/v1/evaluation_data/<int:evaluation_id>")
-        def get_evaluation(evaluation_id):
-            return self.sweep_manager.get_evaluation(evaluation_id)
+        init_app_routes(FlaskHelper.app, self.sweep_manager)
```
Can you move this up so that this happens right after the Flask service is instantiated.

ilmarinen commented on package.json:
```diff
@@ -28,11 +29,21 @@
+    "webpack-dev-server": "^3.11.0"
+  },
+  "scripts": {
+    "start": "webpack-dev-server --hot --mode development",
```
Can you add `--host 0.0.0.0` so that it is not listening on localhost by default.
Recall our conversation regarding my dev setup where its running in a VM with a specific IP so that I'm not running against localhost.

ilmarinen commented on public/index.html:
```diff
@@ -0,0 +1,36 @@
+<!--
```
Can't we re-use the existing Jinja2 template by having the Python app bootstrapping process just render the template as this file?

ilmarinen commented on widget/widget.css:
```diff
@@ -3,6 +3,15 @@
+    font-family: "Helvetica Neue", Helvetica, Arial, sans-serif;
+    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
+ }
+
```
So I did a visual comparison between what the widget looks like in the dev container vs Jupyter, and I still see differences.

Attaching two images. First one is within Jupyter and second one is within the dev window:

<img width="631" alt="Screen Shot 2020-10-14 at 11 09 05 PM" src="https://user-images.githubusercontent.com/680818/96084454-f5a21480-0e73-11eb-8a1a-db342d683d21.png">

<img width="624" alt="Screen Shot 2020-10-14 at 11 21 36 PM" src="https://user-images.githubusercontent.com/680818/96084511-0f435c00-0e74-11eb-8970-bd87158e096a.png">

Notice how some of the spacing is missing?

Nking92 replied:
Thanks for the screenshot. I wasn't sure exactly what else was missing. I'll see if I can fix that.

ilmarinen reviewed (commented):
@Nking92 Looks good. Thank you for making the changes requested so that the development version of the code is almost identical to how we would run the widget within Jupyter. Great job there.

The one lingering point where we differ from the widget is in the fact that we are using two template files. But I suspect that this is needed in order for hot reload to work? If so then I am willing to accept this as a necessary evil. We will just need to remember to keep these files in sync and document that fact somewhere. But if it is not needed then lets try and use just one file.

I am still noticing some differences in rendering between the dev window and within Jupyter. I have attached two images in a comment. Let me know if you are seeing the same thing.

Besides that, I have left a few structural comments regarding where files should live etc. this is minor stuff that needs to be done to keep the structure tidy.

Great job. Thank you!

ilmarinen commented:
Oh one thing that I forgot to mention:

Just an FYI that you can still build with `npx webpack`, you just need to do `NODE_ENV=production npx webpack`.

But in either case, would you mind documenting the steps for:

1. Building the widget and give both alternatives, i.e. `npm run build` and the adjustment I mentioned above.
2. Detailed instructions for running and using the dev harness.

You can add these to the README.


Thank you!

ilmarinen commented on .flaskenv:
```diff
@@ -0,0 +1,2 @@
+FLASK_APP=development/app.py
+FLASK_ENV=development
```
You can add the `FLASK_RUN_PORT=5000` back here, since we're using that in the API and need it to be the same. Othwerwise there may be a time when it chooses some other random port.

ilmarinen commented on README.md:
```diff
@@ -41,6 +41,16 @@ To run tests, make sure that you are in the project root folder and do:
+
+To run the widget and APIs in a development environment:
+1. Open a new terminal, then from within the root folder do `npm start`. This will host the widget with `webpack-dev-server`.
+2. Open a new terminal, then from within the root folder do `flask run`. This will start the Flask server for the APIs used by the widget.
```
Can you also specify options for how one should run this for when you want to develop against an ip other than localhost?

ilmarinen commented on README.md:
```diff
@@ -41,6 +41,16 @@ To run tests, make sure that you are in the project root folder and do:
+1. Open a new terminal, then from within the root folder do `npm start`. This will host the widget with `webpack-dev-server`.
+2. Open a new terminal, then from within the root folder do `flask run`. This will start the Flask server for the APIs used by the widget.
+
+The widget can be loaded in the web browser at `localhost:3000`. It will be loaded independently from a Jupyter notebook. The APIs will be hosted at `localhost:5000`.
```
Some words about the non-localhost case as well should go here.

ilmarinen commented on backwardcompatibilityml/helpers/training.py:
```diff
@@ -827,7 +827,7 @@ def compatibility_sweep(sweeps_folder_path, number_of_epochs, h1, h2,
 
     sweep_summary = {
         "data": sweep_summary_data,
-        "h1_performance": model_accuracy(h1, test_set, device=device)
```
Why is this change here?

ilmarinen reviewed (approved):
@Nking92 +1
